synapse/lib/modelrev.py

```diff
@@ -1205,3 +1206,761 @@ async def _propArrayToForm(self, layers, propfull, formname):
         }
         '''
         await self.runStorm(storm, opts=opts)
+
+class ModelMigration_0_2_31:
+    @classmethod
+    async def anit(cls, core, layers):
+        self = cls()
+
+        self.core = core
+        self.layers = layers
+
+        self.meta = {'time': s_common.now(), 'user': self.core.auth.rootuser.iden}
+
+        self.editcount = 0
+        self.nodeedits = {}
+
+        self.nodes = await s_spooled.Dict.anit(dirn=self.core.dirn)
+        self.todos = await s_spooled.Set.anit(dirn=self.core.dirn)
+
+        self.core.onfini(self.nodes)
+        self.core.onfini(self.todos)
+
+        queues = await self.core.listCoreQueues()
+        queues = {k['name']: k for k in queues}
+
+        if queues.get('model_0_2_31:nodes') is None:
+            await self.core.addCoreQueue('model_0_2_31:nodes', {})
+
+        return self
+
+    async def _queueEdit(self, layriden, edit):
+        self.nodeedits.setdefault(layriden, {})
+        buid, formname, edits = edit
+        self.nodeedits[layriden].setdefault(buid, (buid, formname, []))
+        self.nodeedits[layriden][buid][2].extend(edits)
+        self.editcount += 1
+
+        if self.editcount >= 1000: # pragma: no cover
+            await self._flushEdits()
+
+    async def _flushEdits(self):
+        logger.debug(f'Saving {self.editcount} edits')
+        for layriden, layredits in self.nodeedits.items():
+            layer = self.core.getLayer(layriden)
+            if layer is None: # pragma: no cover
+                continue
+
+            await layer.storNodeEditsNoLift(list(layredits.values()), self.meta)
+
+        self.editcount = 0
+        self.nodeedits = {}
+
+    # NOTE: For the edit* functions below, we only need precise state tracking for nodes and properties. Don't precisely
+    # track the rest.
+    async def editNodeAdd(self, layriden, buid, formname, formvalu, stortype):
+        # logger.debug(f'Adding node: {formname}={repr(formvalu)}, {layriden=}, {buid=}')
+
+        if not self.nodes.has(buid):
+
+            node = {
+                'formname': formname,
+                'formvalu': formvalu,
+                'sodes': {},
+                'nodedata': {},
+                'n1edges': {},
+                'n2edges': {},
+            }
+            await self.nodes.set(buid, node)
+
+        await self._queueEdit(layriden,
+            (buid, formname, (
+                (s_layer.EDIT_NODE_ADD, (formvalu, stortype), ()),
+            )),
+        )
+
+    async def editPropSet(self, layriden, buid, formname, propname, newvalu, oldvalu, stortype):
+        # logger.debug(f'Setting prop: {formname}:{propname}={newvalu} (was {oldvalu})')
+        assert self.nodes.has(buid)
```

Comment:
All of the `assert` statements will be removed when run in containers. Were these checks left in place for development or are they intended to cause the migration to fail if they are violated?

Reply:
Yep, I know they're removed/ignored in release. I left them on purpose for development purposes only.